crates/rover-client/src/shared/lint_response.rs

```diff
@@ -0,0 +1,45 @@
+use prettytable::format::consts::FORMAT_BOX_CHARS;
+use serde::Serialize;
+
+use prettytable::{row, Table};
+use serde_json::{json, Value};
+
+#[derive(Debug, Clone, Serialize, Eq, PartialEq)]
+pub struct LintResponse {
+    pub diagnostics: Vec<Diagnostic>,
+}
+
+impl LintResponse {
+    pub fn get_table(&self) -> String {
+        let mut table = Table::new();
+
+        table.set_format(*FORMAT_BOX_CHARS);
+
+        // bc => sets top row to be bold and center
+        table.add_row(row![bc => "Coordinate", "Line", "Level", "Description"]);
```

Comment:
Since we have nice line/column numbers for the lints, I think we should try a new format that isn't a table (I kinda want to get rid of the table for checks/composition as well but can start here)

```suggestion
        if self.diagnostics.is_empty() {
          // print something like "no changes" - currently this just prints the table heading and nothing else
          format!("No lint violations detected in {filepath}", filepath = &self.filepath)
        } else {
          let mut output = String::new();
          for diagnostic in &self.diagnostics {
            output.push_str(&format!("{level}: In '{coordinate}', {message}, {filepath}:${line}:${column}", level = &diagnostic.level, coordinate = &diagnostic.coordinate, &message = &diagnostic.message, &filepath = &self.filepath, line = &diagnostic.line, col = &diagnostic.column))
          }
          output
        }
```

old output:

```
┌────────────┬──────┬─────────┬────────────────────────────────────────────────────────────────────────┐
│ Coordinate │ Line │  Level  │                              Description                               │
├────────────┼──────┼─────────┼────────────────────────────────────────────────────────────────────────┤
│ Product.id │ 25   │ WARNING │ @tag name “hi-from-products” is not in the list of approved tag names. │
└────────────┴──────┴─────────┴────────────────────────────────────────────────────────────────────────┘

```

new output:

```
WARNING: In 'Product.id', @tag name "hi-from-products" is not in the list of approved tag names, ./schemas/products.graphql:25:0
```

Reply:
Also, I noticed that if `ERROR`s are returned from this list, the lint command does not actually exit with the proper code:

```
$ cargo rover subgraph lint --schema ./examples/supergraph-demo/products/products.graphql --name products averys-federated-graph
    Finished dev [unoptimized + debuginfo] target(s) in 0.49s
     Running `target/debug/rover subgraph lint --schema ./examples/supergraph-demo/products/products.graphql --name products averys-federated-graph`
┌────────────┬──────┬───────┬────────────────────────────────────────────────────────────────────────┐
│ Coordinate │ Line │ Level │                              Description                               │
├────────────┼──────┼───────┼────────────────────────────────────────────────────────────────────────┤
│ Product.id │ 25   │ ERROR │ @tag name “hi-from-products” is not in the list of approved tag names. │
└────────────┴──────┴───────┴────────────────────────────────────────────────────────────────────────┘

$ echo $?
0
```

^ this `echo $?` is returning the exit code of the previous command, which should be `1` in the case that there are any `ERROR`s at all in the response - we want CI pipelines to fail explicitly rather than reporting the errors silently and continuing on to other steps in the pipeline.

Reply:
If you look at `./crates/rover-client/src/shared/check_response.rs`, you'll see that we check the `failure_count` before deciding to return `Ok(check_response` or `Err(RoverClientError::OperationCheckFailure { check_response })`. In the error case, we still report all non-errors as normal, but returning it in the error rather than in the ok means the command will fail with the proper exit code.

Reply:
> Since we have nice line/column numbers for the lints, I think we should try a new format that isn't a table (I kinda want to get rid of the table for checks/composition as well but can start here)

I came in here to say something similar. It'd be really good for our tools to start standardising on diagnostic display. The Router is transitioning its validation to use `apollo-compiler`, which means diagnostics would be displayed using the [`ariadne` crate](https://docs.rs/ariadne/0.2.0/ariadne/). Could the new lint diagnostics in Rover do the same?

I think it'd be nice for our users to see a cohesive diagnostics story throughout their terminal tooling, where they always know how to visually interpret and understand what went wrong

Reply:
here are some examples of diagnostics the compiler produces:

<img width="849" alt="Screenshot 2023-06-02 at 17 34 25" src="https://github.com/apollographql/rover/assets/8107784/5b0be625-875f-4549-be73-a5aea3804f52">
<img width="833" alt="Screenshot 2023-06-02 at 17 37 43" src="https://github.com/apollographql/rover/assets/8107784/1bb5d10a-97ab-408c-9f89-e839b98eeb5a">

<img width="915" alt="Screenshot 2023-06-02 at 17 39 42" src="https://github.com/apollographql/rover/assets/8107784/54714839-5709-42e5-871a-404647304281">

Reply:
I got an example of the linter results printing out using ariadne. The examples on the library were pretty limited so any feedback is welcome. A couple notes I ran into was the byte offset ranges to locate the coordinate in the file all use `usize` so I had to do some conversions from an i64. Also the library does not return a string to be outputed, it does all the printing itself. Curious if there is a better way than how I have integrated it. 


<img width="955" alt="Screenshot 2023-06-07 at 4 23 42 PM" src="https://github.com/apollographql/rover/assets/2107126/d43b5a60-65d9-4dd9-8482-f27aa32d17b9">

Reply:
perhaps [the usage](https://github.com/apollographql/apollo-rs/blob/main/crates/apollo-compiler/src/diagnostics.rs#L84-L147) in `apollo-compiler` might be a better example. But this looks really good already! We can always tweak the display afterwards!

I noticed you grouped all the linter results as a single diagnostic, rather than having them individually like they'd be in clippy or rustc  output. Something we really liked doing in `apollo-compiler` is [having different variants](https://github.com/apollographql/apollo-rs/blob/main/crates/apollo-compiler/src/diagnostics.rs#L138-L360) for our diagnostics corresponding to the validation rules in the spec. It may be good to do that here as well, it may also help with that separation. This could also help with distinguishing between that `warning` and `error`s you have in the screenshot. Perhaps the linting lib already separates out the linting rule types that could map to different diagnostic variants here?

Regardless though, I don't think any of this is a real blocker, and something we can iterate on in the future after this lands! It's super cool to see everything tie together like this!